File: data_structures/doubly_linked_list/cpp/doubly_linked_list.cpp
#include "doubly_linked_list.h"

DoublyLinkedList::~DoublyLinkedList() {
    while (head_) { Node* n = head_->next; delete head_; head_ = n; }
}

void DoublyLinkedList::unlink(Node* n) {
    if (n->prev) n->prev->next = n->next; else head_ = n->next;
    if (n->next) n->next->prev = n->prev; else tail_ = n->prev;
    delete n;
    --size_;
}

void DoublyLinkedList::push_front(int value) {
    Node* n = new Node{value, nullptr, head_};
    if (head_) head_->prev = n; else tail_ = n;
    head_ = n;
    ++size_;
}

void DoublyLinkedList::push_back(int value) {
    Node* n = new Node{value, tail_, nullptr};
    if (tail_) tail_->next = n; else head_ = n;
    tail_ = n;
    ++size_;
}

int DoublyLinkedList::pop_front() {
    if (!head_) throw std::underflow_error("list is empty");
    int v = head_->val;
    unlink(head_);
    return v;
}

int DoublyLinkedList::pop_back() {
    if (!tail_) throw std::underflow_error("list is empty");
    int v = tail_->val;
    unlink(tail_);
    return v;
}

int DoublyLinkedList::front() const {
    if (!head_) throw std::underflow_error("list is empty");
    return head_->val;
}

int DoublyLinkedList::back() const {
    if (!tail_) throw std::underflow_error("list is empty");
    return tail_->val;
}

bool DoublyLinkedList::contains(int value) const {
    for (Node* cur = head_; cur; cur = cur->next)
        if (cur->val == value) return true;
    return false;
}

bool DoublyLinkedList::remove(int value) {
    for (Node* cur = head_; cur; cur = cur->next) {
        if (cur->val == value) { unlink(cur); return true; }
    }
    return false;
}

int DoublyLinkedList::size() const { return size_; }
bool DoublyLinkedList::is_empty() const { return size_ == 0; }

std::vector<int> DoublyLinkedList::to_vector() const {
    std::vector<int> v;
    for (Node* cur = head_; cur; cur = cur->next) v.push_back(cur->val);
    return v;
}

File: data_structures/doubly_linked_list/python/doubly_linked_list.py
"""双方向連結リスト

操作:
  push_front(v) / push_back(v) : O(1)
  pop_front() / pop_back()     : O(1)
  front() / back()             : O(1)
  contains(v) / remove(v)      : O(n)
  size() / is_empty()          : O(1)
  to_list()                    : O(n)
"""

from typing import List, Optional


class _Node:
    __slots__ = ("val", "prev", "next")

    def __init__(self, val: int) -> None:
        self.val = val
        self.prev: Optional["_Node"] = None
        self.next: Optional["_Node"] = None


class DoublyLinkedList:
    def __init__(self) -> None:
        self._head: Optional[_Node] = None
        self._tail: Optional[_Node] = None
        self._size = 0

    def _unlink(self, node: _Node) -> None:
        if node.prev:
            node.prev.next = node.next
        else:
            self._head = node.next
        if node.next:
            node.next.prev = node.prev
        else:
            self._tail = node.prev
        self._size -= 1

    def push_front(self, value: int) -> None:
        node = _Node(value)
        node.next = self._head
        if self._head:
            self._head.prev = node
        else:
            self._tail = node
        self._head = node
        self._size += 1

    def push_back(self, value: int) -> None:
        node = _Node(value)
        node.prev = self._tail
        if self._tail:
            self._tail.next = node
        else:
            self._head = node
        self._tail = node
        self._size += 1

    def pop_front(self) -> int:
        if self._head is None:
            raise IndexError("list is empty")
        v = self._head.val
        self._unlink(self._head)
        return v

    def pop_back(self) -> int:
        if self._tail is None:
            raise IndexError("list is empty")
        v = self._tail.val
        self._unlink(self._tail)
        return v

    def front(self) -> int:
        if self._head is None:
            raise IndexError("list is empty")
        return self._head.val

    def back(self) -> int:
        if self._tail is None:
            raise IndexError("list is empty")
        return self._tail.val

    def contains(self, value: int) -> bool:
        cur = self._head
        while cur:
            if cur.val == value:
                return True
            cur = cur.next
        return False

    def remove(self, value: int) -> bool:
        cur = self._head
        while cur:
            if cur.val == value:
                self._unlink(cur)
                return True
            cur = cur.next
        return False

    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    def to_list(self) -> List[int]:
        result = []
        cur = self._head
        while cur:
            result.append(cur.val)
            cur = cur.next
        return result
